File: day2/lightbulbs/submissions/partially_accepted/jb_staircase_2n.py
#!/bin/python3

def query(q, t='?'):
  print(t)
  grid = [['0' for i in range(n)] for i in range(n)]
  for r,c in q:
    grid[r-1][c-1] = '1'
  for row in grid:
    print("".join(row))
  if t == '!':
    exit()
  return int(input())

n = int(input())

h, v = [], []

a = query([(1,1), (1,2), (1,3)])
if a == 3*n: v.append((1,1))
elif a == n: h.append((1,1))
elif a-query([(1,2), (1,3)]) == n-1: v.append((1,1))
else: h.append((1,1))

while len(h) < n and len(v) < n:
  p = (1+len(h), 1+len(v))
  (hh,vv) = (h,v) if len(h) else (v,h)
  if query(hh+[p]) == (len(hh)+1)*n:
    hh.append(p)
  else:
    vv.append(p)

query(h if len(h) == n else v,'!')

File: day2/lightbulbs/submissions/partially_accepted/jb_staircase_2n.cc
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for (int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
#define debug(...) // ignore
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
typedef long double ld;

int main() {
  int n;
  cin >> n;

  bool transpose = false;
  auto printgrid = [&](vector<pii> v) {
    if (transpose) for (auto &[a, b] : v) swap(a, b);
    vector<string> grid(n,string(n,'0'));
    for(auto [a,b] : v) grid[a][b] = '1';
    for(auto g : grid) cout << g << endl;
  };
  auto query = [&](vector<pii> v) {
    cout << "?" << endl;
    printgrid(v);
    int x;
    cin >> x;
    return x;
  };
  auto answer = [&](vector<pii> v) {
    assert(sz(v) == n);
    cout << "!" << endl;
    printgrid(v);
    exit(0);
  };

  {
    int a = query({{0, 0}, {0, 1}, {0, 2}});
    int b = query({{0, 1}, {0, 2}});
    set<pii> vert = {{3 * n, 2 * n}, {3 * n - 2, 2 * n - 1}, {2 * n - 1, n}};
    set<pii> horz = {{3 * n - 2, 2 * n}, {2 * n - 1, 2 * n - 1}, {n, n}};
    if (vert.count(pii(a, b))) transpose = true;
    else assert(horz.count(pii(a, b)));
    // now assume top-left is horizontal
  }
  vector<pii> h = {{0, 0}}, v;
  int x = 1, y = 0;
  while (x < n && y < n) {
    int q = query({{0, 0}, {x, y}});
    if (q == 2 * n) h.emplace_back(x++, y);
    else if (q == 2 * n - 1) v.emplace_back(x, y++);
    else assert(false);
  }
  if (x == n) answer(h);
  else answer(v);
}
